File: lab/1/linear_regression.py
import random
import math
from tensor import Tensor

# y = a * x + b
real_a = 3.14
real_b = -17.99
# generate dataset
N = 100
Ys = []
Xs = []
for i in range(N):
  x = (random.random() - 0.5) * 2  # (-1, 1)
  y = real_a * x + real_b
  y += (random.random() - 0.5) * 0.01 # add random error
  Xs.append(x)
  Ys.append(y)

# learning
epochs = 100
lr = 0.001
A = Tensor(random.random())
B = Tensor(random.random())
for epoch in range(epochs):
  epoch_loss = 0.0
  for i in range(N):
    # forward pass, compute the loss
    x = Xs[i]
    y_true = Ys[i]
    X = Tensor(x)
    Y_TRUE = Tensor(y_true)
    Y_PRED = A * X + B
    LOSS = (Y_TRUE - Y_PRED) * (Y_TRUE - Y_PRED)
    epoch_loss += LOSS.value
    # backward pass, compute the LOSS gradients w.r.t A and B
    LOSS.reset_gradient()
    LOSS.backward()
    # apply SGD
    A.value -= lr * A.gradient
    B.value -= lr * B.gradient

  print(f'epoch={epoch}/{epochs}, loss={epoch_loss/N}, A={A.value}, B={B.value}')


print(f'real_a={real_a}, real_b={real_b}')
print(f'learned_a={A.value:.2f}, learned_b={B.value:.2f}')

assert math.isclose(A.value, real_a, rel_tol=1e-02) # Python 3.5 or newer
assert math.isclose(B.value, real_b, rel_tol=1e-02)

File: lab/2/linear_regression.cc
#include "lab/2/tensor.h"
#include <random>
#include <glog/logging.h>

using zeronn::lab2::Tensor;

const float kRealA = 3.14;
const float kRealB = -17.99;

void GenerateDataset(std::vector<float>& Xs, std::vector<float>& Ys, int N) {
  Xs.resize(N);
  Ys.resize(N);
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(-1.0, 1.0);
  for (int i = 0; i < N; ++i) {
    float x = dis(gen);
    float y = kRealA * x + kRealB;
    y += dis(gen) * 0.005;
    Xs[i] = x;
    Ys[i] = y;
  }
}

int main() {
  // generate dataset
  const int N = 100;
  std::vector<float> Xs, Ys;
  GenerateDataset(Xs, Ys, N);
  // learning
  const int epochs = 100;
  const float lr = 0.001;
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(-1.0, 1.0);
  Tensor A(dis(gen));
  Tensor B(dis(gen));
  for (int epoch = 0; epoch < epochs; ++epoch) {
    float epoch_loss = 0.0;
    for (int i = 0; i < N; ++i) {
      // forward pass
      float x = Xs[i];
      float y_true = Ys[i];
      Tensor X(x);
      Tensor Y_TRUE(y_true);
      auto Y_PRED = A * X + B;
      auto LOSS = (Y_TRUE - Y_PRED) * (Y_TRUE - Y_PRED);
      epoch_loss += LOSS.data_;
      // backward pass
      LOSS.ResetGradient();
      LOSS.Backward();
      // SGD
      A.data_ -= lr * A.gradient_;
      B.data_ -= lr * B.gradient_;
      // clean up
      Tensor::ReleaseIntermediateTensors();
    }
    LOG(INFO) << "epoch=" << epoch << "/" << epochs << ", loss=" << epoch_loss / N << ", A=" << A.data_ << ", B=" << B.data_;
  }

  LOG(INFO) << "kRealA=" << kRealA << ", kRealB=" << kRealB;
  LOG(INFO) << "learned A=" << A.data_ << ", learned B=" << B.data_;

  return 0;
}

